### PR TITLE
implement additional driver.Rows interfaces

### PR DESCRIPTION
**Pull request by hugorut**
Currently type `ocRows` doesn't implement the additional `driver.Rows` interfaces. 

* `driver.RowsNextResultSet`
* `driver.RowsColumnTypePrecisionScale`
* `driver.RowsColumnTypeNullable`
* `driver.RowsColumnTypeLength`
* `driver.RowsColumnTypeDatabaseTypeName`

This causes issues in available public methods such as `sql.Rows.ColumnTypes` which does explicit interface checking to populate results:

```go
func rowsColumnInfoSetupConnLocked(rowsi driver.Rows) []*ColumnType {
    ...
    if prop, ok := rowsi.(driver.RowsColumnTypeScanType); ok {
        ci.scanType = prop.ColumnTypeScanType(i)
    }
    ...
}
```

This PR fixes this issue by implementing the necessary interface methods.

### CODE REVIEW

File: driver.go [hunk 1]
```diff
@@ -604,6 +605,83 @@ type ocRows struct {
...
+
+// NextResultsSet calls the implements the driver.RowsNextResultSet for ocRows.
+// It returns the the underlying result of NextResultSet from the ocRows.parent
+// if the parent implements driver.RowsNextResultSet.
+func (r ocRows) NextResultSet() error {
+	if v, ok := r.parent.(driver.RowsNextResultSet); ok {
+		return v.NextResultSet()
+	}
+
+	return nil
```
Comment (basvanbeek) [inline driver.go]: `driver.go` states "NextResultSet should return io.EOF when there are no more result sets."

Probably it's better to send `io.EOF` if underlying rows implementation does not support driver.RowsNextResultSet

File: driver.go [hunk 2]
```diff
@@ -604,6 +605,83 @@ type ocRows struct {
...
+}
+
+// ColumnTypePrecisionScale calls the implements the driver.RowsColumnTypePrecisionScale for ocRows.
```
Comment (basvanbeek) [inline driver.go]: Please move around these method implementations to follow the same order as their interface definitions in the driver package.

File: driver.go [hunk 3]
```diff
@@ -604,6 +605,83 @@ type ocRows struct {
...
+
+// ColumnTypeScanType calls the implements the driver.RowsColumnTypeScanType for ocRows.
+// It returns the the underlying result of ColumnTypeScanType from the ocRows.parent
+// if the parent implements driver.RowsColumnTypeScanType.
+func (r ocRows) ColumnTypeScanType(index int) reflect.Type {
+	if v, ok := r.parent.(driver.RowsColumnTypeScanType); ok {
+		return v.ColumnTypeScanType(index)
+	}
+
+	return nil
```
Comment (basvanbeek) [inline driver.go]: This method implementation is not transparent. The code actually using this interface does the following:

```go
if prop, ok := rowsi.(driver.RowsColumnTypeScanType); ok {
	ci.scanType = prop.ColumnTypeScanType(i)
} else {
	ci.scanType = reflect.TypeOf(new(interface{})).Elem()
}
```

So either return `reflect.TypeOf(new(interface{})).Elem()` (brittle as the go sql code using this functionality might change)  or hide this interface extension when wrapped drivers do not implement this interface

**Review by basvanbeek — CHANGES_REQUESTED** [review]
Some minor nits to iron out and then it looks ok to me.

**Comment by hugorut** [comment]
@basvanbeek changes made + addition of tests

**Comment by hugorut** [comment]
travis build seemed to be failing because of a recent patch to the golint library. It now enforces that you import using:

```
go get -u golang.org/x/lint/golint 
```

see issue for more details: https://github.com/golang/lint/issues/415

Have updated the travis script to reflect.

**Comment by hugorut** [comment]
@basvanbeek any chance we can get this merged?

**Comment by basvanbeek** [comment]
@hugorut thank you for the contribution, greatly appreciated. I'd like a couple of changes for consistency's sake:

We have enhancement/optional interface logic happening for `ocDriver` and `ocConn` too. These have been handled in a specific way by having the underlying implementation implement all optional interface methods and then at wrapping time inspect which interfaces are supported by the underlying driver. From there the actual implementation is wrapped by structs enabling only the supported interfaces or if all interfaces are supported the implementation is returned unwrapped. 

The functions handling the wrapping are called `wrapDriver` and `wrapConn`. So in this case `newRows` would become `wrapRows` and should handle the logic for the Rows interfaces the same way.

The majority of your test suite can then also be removed because we can compile time assert the `ocRows` implementation against the interfaces.

It would be great to have proper tests for all the wrap functions (testing each combination of interfaces being correctly opened up) but this is similarly brittle as the actual wrapping code if written by hand.

I expect these tests to be added in the future but they should be auto generated.

**Comment by hugorut** [comment]
@basvanbeek sorry for the radio silence on this. Snowed with other things.

Apologies that I did not see the `wrap` functions before. However, the implementation within those functions will not work for the driver.Rows additional interfaces :(.

The current implementation in `wrapStmt` (for example) works because the underlying additional interfaces are isolated. e.g. `driver.tmtExecContext` contains a single method. However with the `driver.Rows` interfaces they *compose* the interface from the parent `driver.Rows` interface. e.g:

```go
type RowsColumnTypeScanType interface {
	Rows
	ColumnTypeScanType(index int) reflect.Type
}
```

hence if we do something like this:

```go
return struct {
       driver.Rows
       driver.RowsColumnTypeScanType
}{rs, rcts}
``` 

go compilation will error with:

```bash
struct { driver.Rows; driver.RowsColumnTypeScanType }.Close is ambiguous
```

The solution is to create concrete struct types for each interface combination. However this feels like overkill. Let me know your thoughts.

**Comment by basvanbeek** [comment]
No problem, we all get busy. Yeah I've noticed the composition too. Not sure why they did that but we have to live with it. We can however declare our own local (even unexported) interfaces and use those to achieve the desired effect. Implicit interfaces FTW.

**Comment by hugorut** [comment]
Cool perfect was going to suggest such an approach. Will implement this afternoon.

**Comment by basvanbeek** [comment]
Awesome, thank you for hanging in there. Transparency of this wrapper without introducing side effects is super important to me.

**Comment by hugorut** [comment]
So I just realised if we take the switch composition approach we're going to have to write 64 combinations of struct initialisation. Not only will this be cumbersome but also could be open to a lot of user error. I'm kinda unsure of the best way to achieve what we want without simply going back to the `ocRows` implementing a generic version of the additional interfaces which calls the parent if it's implemented.  Thoughts?

**Comment by basvanbeek** [comment]
Fortunately not needed. We only have two compositions that matter. One with and one without `driver.RowsColumnTypeScanType`. The others are safe to have return default values if the underlying driver does not support them.

**Comment by hugorut** [comment]
Ok phew great. I thought I was going to have to do that for every branch

**Comment by basvanbeek** [comment]
@hugorut are you ok with me closing this PR in favor of continuing in PR #26?

**Comment by hugorut** [comment]
I'll close now #26 is in play

**Pull request closed without merging**